solver/llbsolver/history.go

```diff
@@ -17,24 +30,117 @@ const (
 type HistoryQueueOpt struct {
 	DB           *bolt.DB
 	LeaseManager leases.Manager
+	ContentStore content.Store
+	CleanConfig  *config.HistoryConfig
 }
 
 type HistoryQueue struct {
 	mu       sync.Mutex
 	initOnce sync.Once
 	HistoryQueueOpt
-	ps     *pubsub[*controlapi.BuildHistoryEvent]
-	active map[string]*controlapi.BuildHistoryRecord
+	ps      *pubsub[*controlapi.BuildHistoryEvent]
+	active  map[string]*controlapi.BuildHistoryRecord
+	refs    map[string]int
+	deleted map[string]struct{}
 }
 
 func NewHistoryQueue(opt HistoryQueueOpt) *HistoryQueue {
-	return &HistoryQueue{
+	if opt.CleanConfig == nil {
+		opt.CleanConfig = &config.HistoryConfig{
+			MaxAge:     int64((48 * time.Hour).Seconds()),
+			MaxEntries: 50,
+		}
+	}
+	h := &HistoryQueue{
 		HistoryQueueOpt: opt,
 		ps: &pubsub[*controlapi.BuildHistoryEvent]{
 			m: map[*channel[*controlapi.BuildHistoryEvent]]struct{}{},
 		},
-		active: map[string]*controlapi.BuildHistoryRecord{},
+		active:  map[string]*controlapi.BuildHistoryRecord{},
+		refs:    map[string]int{},
+		deleted: map[string]struct{}{},
+	}
+
+	go func() {
+		for {
+			h.gc()
+			time.Sleep(120 * time.Second)
+		}
+	}()
+
+	return h
+}
+
+func (h *HistoryQueue) gc() error {
+	var records []*controlapi.BuildHistoryRecord
+
+	if err := h.DB.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(recordsBucket))
+		if b == nil {
+			return nil
+		}
+		return b.ForEach(func(key, dt []byte) error {
+			var br controlapi.BuildHistoryRecord
+			if err := br.Unmarshal(dt); err != nil {
+				return errors.Wrapf(err, "failed to unmarshal build record %s", key)
+			}
+			if br.Pinned {
+				return nil
+			}
+			records = append(records, &br)
+			return nil
+		})
+	}); err != nil {
+		return err
+	}
+
+	if len(records) < int(h.CleanConfig.MaxEntries) {
+		return nil
+	}
+
+	sort.Slice(records, func(i, j int) bool {
+		return records[i].CompletedAt.Before(*records[j].CompletedAt)
+	})
```

Comment:
I'm confused between the interaction of `MaxAge` and `MaxEntries` here. I would expect that records be cleaned if it was either over the max age, or if the number of entries was exceeded.

I think in reading below as well, there a scenario in which we don't delete all everything past `MaxEntries`, so we could finish the gc loop with more than `MaxEntries`?

Reply:
Added comment